### PR TITLE
[pytorch] [build, test] Adding pt support for horovod and MPI/OpenSSH and include Horovod Tests

### PR DESCRIPTION
**Pull request by harshp8l**
*Issue #, if available:*

## Checklist
- [ ] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [build] | [test] | [build, test] | [ec2, ecs, eks, sagemaker]
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

*Description:*
Currently, this includes the config necessary for openssh/mpi
This PR is pending tests from the SageMaker PyTorch Estimator API for including horovod.

*Tests run:*

*DLC image/dockerfile:*

*Additional context:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: pytorch/training/docker/1.5.0/py3/Dockerfile.cpu [hunk 1]
```diff
@@ -106,6 +106,16 @@ RUN pip install --upgrade pip --trusted-host pypi.org --trusted-host files.pytho
  && pip install --no-deps --no-cache-dir -U \
     https://torchvision-build.s3.amazonaws.com/1.5.0/cpu/torchvision-0.6.0-cp36-cp36m-linux_x86_64.whl
 
+# Adding support to SageMaker for MPI/Horovod
```
Comment (gautamkmr) [inline pytorch/training/docker/1.5.0/py3/Dockerfile.cpu:109]: What about 1.5.1 ?

Comment (harshp8l) [reply pytorch/training/docker/1.5.0/py3/Dockerfile.cpu:109]: Rest of work here is dependent on SageMaker ML Frameworks Team changes, the latest they support as of now is 1.5.0 (https://github.com/aws/sagemaker-pytorch-training-toolkit/tree/master/docker).

**Comment by arjkesh** [comment]
Hi @harshp8l -- is this still WIP?

**Comment by harshp8l** [comment]
@bveeramani Will be providing the necessary updates here

**Comment by TusharKanekiDey** [comment]
@bveeramani Any updates on this PR ?

**Comment by TusharKanekiDey** [comment]
Do we need similar changes for PT 1.5.1 and PT1.6.0

**Comment by bveeramani** [comment]
> @bveeramani Any updates on this PR ?

Sorry for the late reply, I hadn't seen this message. I've been blocked by so minor errors, but I'm spending all of the next several days working on fishing this

**Comment by bveeramani** [comment]
> Do we need similar changes for PT 1.5.1 and PT1.6.0 ?

I would think so yes, so that users can use Horovod with those framework versions

**Comment by bveeramani** [comment]
Update: for the Horovod SageMaker tests to run correctly, the PyTorch Training Toolkit needs to be updated to use MPI.

See aws/sagemaker-pytorch-training-toolkit#217.

**Comment by bveeramani** [comment]
Update: wrote up an issue describing what work remains for this PR. https://github.com/aws/deep-learning-containers/issues/553

**Comment by ChaiBapchya** [comment]
Any update on this PR? @saimidu @harshp8l @arjkesh @bveeramani  Can we get this merged?

**Comment by harshp8l** [comment]
I had no further updates; not sure on the status of the remaining work items left by @bveeramani

File: pytorch/training/docker/1.6.0/py3/cu101/Dockerfile.gpu [hunk 2]
```diff
@@ -172,6 +173,19 @@ RUN apt-get update \
...
+  sed 's@session\s*required\s*pam_loginuid.so@session optional pam_loginuid.so@g' -i /etc/pam.d/sshd
+
+RUN rm -rf /root/.ssh/ && \
+ mkdir -p /root/.ssh/ && \
+ ssh-keygen -q -t rsa -N '' -f /root/.ssh/id_rsa && \
+ cp /root/.ssh/id_rsa.pub /root/.ssh/authorized_keys \
+ && printf "Host *\n StrictHostKeyChecking no\n" >> /root/.ssh/config
+
+# Comment out line in MPI config to prevent mutually exclusive MCA settings
+RUN sed -i '62,62 s/^/#/' /home/.openmpi/etc/openmpi-mca-params.conf
```
Comment (saimidu) [inline pytorch/training/docker/1.6.0/py3/cu101/Dockerfile.gpu:187]: Is there a cleaner way to do this?

**Review by saimidu — COMMENTED** [review]
LGTM

File: pytorch/training/docker/1.6.0/py3/cu101/Dockerfile.gpu (hunk 2 above)
Comment (ChaiBapchya) [reply pytorch/training/docker/1.6.0/py3/cu101/Dockerfile.gpu:187]: to just remove the lines that add those openmpi mca params?

Comment (saimidu) [reply pytorch/training/docker/1.6.0/py3/cu101/Dockerfile.gpu:187]: Yeah

File: pytorch/training/docker/1.6.0/py3/cu101/Dockerfile.gpu [hunk 3]
```diff
@@ -145,6 +145,7 @@ RUN pip uninstall -y horovod \
  && ldconfig /usr/local/cuda-10.1/targets/x86_64-linux/lib/stubs \
  && HOROVOD_GPU_ALLREDUCE=NCCL HOROVOD_CUDA_HOME=/usr/local/cuda-10.1 HOROVOD_WITH_PYTORCH=1 pip install --no-cache-dir horovod==${HOROVOD_VERSION} \
  && ldconfig
+ && pip install mpi4py==3.0.3
```
Comment (saimidu) [inline pytorch/training/docker/1.6.0/py3/cu101/Dockerfile.gpu:148]: `&& ldconfig` requires ` \` after it for the rest of the multi-line RUN statement to work.

Comment (harshp8l) [reply pytorch/training/docker/1.6.0/py3/cu101/Dockerfile.gpu]: Addressed this change across all the dockerfiles

File: test/sagemaker_tests/pytorch/training/integration/sagemaker/test_horovod.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+from ...integration import resources_path, DEFAULT_TIMEOUT
+from ...integration.sagemaker.timeout import timeout
+
+
+@pytest.mark.processor("gpu")
+@pytest.mark.multinode(2)
+@pytest.mark.integration("horovod")
+@pytest.mark.skip_cpu
+def test_horovod_simple(
```
Comment (saimidu) [inline test/sagemaker_tests/pytorch/training/integration/sagemaker/test_horovod.py:30]: Please add the decorator `@pytest.mark.model("mnist")` to denote that it uses an MNIST model for the training test. This is necessary for it to pass the coverage test on the CI.

File: test/sagemaker_tests/pytorch/training/integration/local/test_horovod.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+import pytest
+from sagemaker.pytorch import PyTorch
+
+from ...integration import resources_path
+
+
+@pytest.mark.integration("horovod")
+@pytest.mark.processor("gpu")
+@pytest.mark.skip_cpu
+def test_horovod_simple(sagemaker_local_session, docker_image, tmpdir):
```
Comment (saimidu) [inline test/sagemaker_tests/pytorch/training/integration/local/test_horovod.py:28]: Please add the decorator `@pytest.mark.model("mnist")` to denote that it uses an MNIST model for the training test. This is necessary for it to pass the coverage test on the CI.

File: test/sagemaker_tests/pytorch/training/integration/local/test_horovod.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+            actual = json.load(file)
+        expected = {'local-rank': local_rank, 'rank': rank, 'size': size}
+
+        assert actual == expected
+
+
+@pytest.mark.integration("horovod")
+@pytest.mark.processor("gpu")
+@pytest.mark.skip_cpu
+def test_horovod_training(sagemaker_local_session, docker_image, tmpdir):
```
Comment (saimidu) [inline test/sagemaker_tests/pytorch/training/integration/local/test_horovod.py:65]: Please add the decorator `@pytest.mark.model("mnist")` to denote that it uses an MNIST model for the training test. This is necessary for it to pass the coverage test on the CI.

File: test/sagemaker_tests/pytorch/training/integration/sagemaker/test_horovod.py [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+        expected = {"local-rank": local_rank, "rank": rank, "size": size}
+
+        assert actual == expected
+
+
+@pytest.mark.processor("gpu")
+@pytest.mark.multinode(2)
+@pytest.mark.integration("horovod")
+@pytest.mark.skip_cpu
+def test_horovod_training(
```
Comment (saimidu) [inline test/sagemaker_tests/pytorch/training/integration/sagemaker/test_horovod.py:84]: Please add the decorator `@pytest.mark.model("mnist")` to denote that it uses an MNIST model for the training test. This is necessary for it to pass the coverage test on the CI.

**Comment by SergTogul** [comment]
@ harshp8l Could you address test failures?

**Comment by arjkesh** [comment]
This PR hasn't been touched in awhile so we are closing this out. Please reopen as needed.

**Pull request closed without merging**